lib/rust/ensogl/core/src/display/navigation/navigator.rs

```diff
@@ -102,17 +125,25 @@ impl NavigatorModel {
             let z              = half_height / camera.half_fovy_slope();
             let direction      = Vector3(x,y,z).normalize();
             let mut position   = simulator.target_value();
-            let min_zoom       = camera.clipping().near + min_zoom;
             let zoom_amount    = zoom.amount * position.z;
-            let direction      = direction   * zoom_amount;
-            let max_zoom_limit = max_zoom - position.z;
-            let min_zoom_limit = min_zoom - position.z;
-            let too_far        = direction.z > max_zoom_limit;
-            let too_close      = direction.z < min_zoom_limit;
-            let zoom_factor    = if too_far   { max_zoom_limit / direction.z }
-                            else if too_close { min_zoom_limit / direction.z }
-                            else              { 1.0 };
-            position          += direction * zoom_factor;
+            let translation    = direction   * zoom_amount;
+
+            // Usage of min/max prefixes might be confusing here. Remember that the max zoom means
+            // the maximum visible size, thus the minimal distance from the camera and vice versa.
+            let fovy_slope               = camera.half_fovy_slope();
+            let distance_to_show_full_ui = scene.shape().value().height / 2.0 / fovy_slope;
+            let max_zoom                 = max_zoom_limit.get().unwrap_or(max_zoom);
+            let min_distance =  1.0 / max_zoom * distance_to_show_full_ui + camera.clipping().near;
+            let max_distance = (1.0 / min_zoom * distance_to_show_full_ui).min(camera.clipping().far);
+
+            let max_translation_limit = max_distance - position.z;
+            let min_translation_limit = min_distance - position.z;
+            let too_far               = translation.z > max_translation_limit;
+            let too_close             = translation.z < min_translation_limit;
+            let limiting_factor       = if too_far   { max_translation_limit / translation.z }
+                                        else if too_close { min_translation_limit / translation.z }
+                                        else              { 1.0 };
+            position                  += translation * limiting_factor;
             simulator.set_target_value(position);
```

Comment:
I see several issues with this code:
1. The comment about min/max prefixes is added to a `fovy_slope` var, which does not make sense IMO.
2. The code is more repetitive than it was. There are lines which are just copy-paste from a few lines above.
3. Some variables are really poorly named, like `let max_zoom = max_zoom_limit.get().unwrap_or(max_zoom);`. So `max_zoom` is either `max_zoom_limit` or `max_zoom`? If you already have `max_zoom`, why are you checking the limit? I don't understand the logic of the code, probably because of var names.
4. There are other strange names, like `limiting_factor` - why `limiting_factor` multiplied by `translation` gives us position? Of course I like the name `translation` way better than `direction` in the old code.
5. Another strange name is `distance_to_show_full_ui` - I know it was in previous code, probably someone did not catch it during review – EnsoGL is a rendering framework for vector shapes – what is `ui` there? There is no UI concept in EnsoGL core.
6. Again var naming problem `max_translation_limit` - earlier you are using `translation` for 3D value. Now, you define `max_translation_limit` as `max_distance - position.z` which seems like 1D value or a number. So the naming is confusing.
7. I do not understand where the equations for `min_distance` and `max_distance` come from. Please refactor the to really well named variables that explain the concepts of things that are happening here. Moreover, I feel that these equations are not correct. Why you have `1/max_zoom` there? Anyway, these equations should be explained and should be refactored to well-named variables.
8. Several concepts are strange there, like `translation` refers to 3D variable, so writing `translation.z > max_translation_limit` does not make sense. In previous code `direction.z > max_zoom_limit` that was OK, as `zoom` is a number, `direction` is 3D variable, so you can compare `direction.z` to a number.
8. The code is not aligned – the if-then-else block is messy.

Reply:
1. Fair point, fixed it. The comment was left for the block of code (separated by blank lines) which is responsible for a distinct part of the computation.
2. Fixed, by introducing a closure in scope of the function. There are no places in the codebase where this equation is used, as far as I'm aware of.
3. This is a common pattern in Rust codebases, to shadow variable name with a new binding applying some additional operations on it. However, I slightly refactored that part of the code, as Adam suggested in his review.
4. I don't understand why `zoom_factor` is a better name than `limiting_factor`. We *limit* the movements of the camera near min/max distance along the z-axis. I've left it as it is (added a small comment), please suggest a better name if you want to change it. `zoom_factor` is extremely confusing, because we already had `zoom_amount` and we've already multiplied it by direction/translation.
5. Makes sense, however, it's hard to come up with a short name here. I choose `distance_to_zoom_factor_of_1` which is rather long but looks self-explanatory to me. Please suggest a better name if you have an idea.
6. It's not clear to me, what other variable names I should revisit. You covered every one of them in this message, with an exception of `too_far` and `too_close`. Please ping me if you want to change anything you didn't explicitly mention. I renamed `max(min)_translation_limit` to `positive(negative)_z_translation_limit`. However, I don't agree that `direction.z > max_zoom_limit` was any better, mainly because `max_zoom_limit` was representing distance rather than actual zoom limit.
7. You're right, it doesn't make much sense from math standpoint, I was playing with different equations while trying to understand this code, and completely missed that `1.0 / x * y` can obviously be replaced with `y / x`. I changed it. I'm not sure I should (and I can) explain why `min_distance = distance_to_zoom_factor_of_1 / max_zoom` - it's just how zooming works, we have a proportion of `1.0 / distance_1 = max_zoom / min_distance`.
8. Fair point, however, we do not have such requirements in our Code Style guide. I've left a comment about it in https://github.com/enso-org/enso/pull/3172